cpp/src/arrow/compute/kernels/vector_cumulative_ops_test.cc

```diff
@@ -0,0 +1,265 @@
+  using ArrowType = typename TestCumulativeOp<T, OptionsType>::ArrowType;
+  using ScalarType = typename TestCumulativeOp<T, OptionsType>::ScalarType;
+  using CType = typename TestCumulativeOp<T, OptionsType>::CType;
+
+ protected:
+  template <typename U = T>
+  enable_if_parameter_free<U, OptionsType> generate_options(CType start = 0,
+                                                            bool skip_nulls = false,
+                                                            bool check_overflow = false) {
+    return OptionsType(std::make_shared<ScalarType>(start), skip_nulls, check_overflow);
+  }
+
+  template <typename U = T>
+  enable_if_t<is_time_type<U>::value || is_timestamp_type<U>::value, OptionsType>
+  generate_options(CType start = 0, bool skip_nulls = false,
+                   bool check_overflow = false) {
+    TimeUnit::type unit;
+    switch (ArrowType::type_id) {
+      case Type::TIME64:
+        unit = TimeUnit::NANO;
+        break;
+      default:
+        unit = TimeUnit::SECOND;
+        break;
+    }
+    return OptionsType(std::make_shared<ScalarType>(start, unit), skip_nulls,
+                       check_overflow);
+  }
+
+  void Assert(const std::string& values, const std::string& expected,
+              const OptionsType& options) {
+    auto values_arr = TestCumulativeOp<T, OptionsType>::array(values);
+    auto expected_arr = TestCumulativeOp<T, OptionsType>::array(expected);
+    auto func_name = options.check_overflow ? "cumulative_sum_checked" : "cumulative_sum";
+    TestCumulativeOp<T, OptionsType>::Assert(func_name, values_arr, expected_arr,
+                                             options);
+  }
+
+  void Assert(const std::vector<std::string>& values,
+              const std::vector<std::string>& expected, const OptionsType& options) {
+    auto values_arr = TestCumulativeOp<T, OptionsType>::chunked_array(values);
+    auto expected_arr = TestCumulativeOp<T, OptionsType>::chunked_array(expected);
+    auto func_name = options.check_overflow ? "cumulative_sum_checked" : "cumulative_sum";
+    TestCumulativeOp<T, OptionsType>::Assert(func_name, values_arr, expected_arr,
+                                             options);
+  }
+};
+
+TYPED_TEST_SUITE(TestCumulativeSum, NumericArrowTypes);
+
+TYPED_TEST(TestCumulativeSum, NoStartNoSkipNoNulls) {
+  CumulativeSumOptions options = this->generate_options();
+  auto empty = "[]";
+  auto values = "[1, 2, 3, 4, 5, 6]";
+  auto expected = "[1, 3, 6, 10, 15, 21]";
+  std::vector<std::string> chunked_values = {"[1, 2, 3]", "[4, 5, 6]"};
+  std::vector<std::string> chunked_expected = {"[1, 3, 6, 10, 15, 21]"};
+
+  this->Assert(empty, empty, options);
+  this->Assert(values, expected, options);
+  this->Assert(chunked_values, chunked_expected, options);
+}
+
+TYPED_TEST(TestCumulativeSum, NoStartNoSkipHasNulls) {
+  CumulativeSumOptions options = this->generate_options();
+  auto one_null = "[null]";
+  auto three_null = "[null, null, null]";
```

Comment:
Nit: I think you use either `one_null` or `three_null` but not both. Similar for all other cases.